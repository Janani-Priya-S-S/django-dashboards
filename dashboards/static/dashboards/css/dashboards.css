:root {
  --color-background-primary: #e5e5e5;
  --color-background-secondary: #fafafa;
  --color-background-tertiary: #f5f5f5;
  --color-border: #d4d4d4;
  --color-text-primary: #404040;
  --color-text-secondary: #a3a3a3;
  --shadow: 0 0 0 0px #fff, 0 20px 25px -5px rgba(0,0,0,0.10196), 0 8px 10px -6px rgba(0,0,0,0.10196);
  --root-background: linear-gradient(to top right, var(--color-background-primary), var(--color-background-primary), var(--color-border));
  --color-danger: #dc2626;
  --color-cta-background: #737373;
  --color-cta-hover-background: #404040;
  --color-cta-text: #fafafa;
  --color-cta-hover-text: var(--color-cta-text);
  --color-cta-alt-background: #a3a3a3;
  --color-cta-alt-hover-background: #fafafa;
  --color-cta-alt-text: #404040;
  --color-cta-alt-hover-text: var(--color-cta-alt-text);
  --transition-time: 0.25s;
}

@media (prefers-color-scheme: dark) {
  :root {
    --color-background-primary: #404040;
    --color-background-secondary: #525252;
    --color-background-tertiary: #737373;
    --color-border: #737373;
    --color-text-primary: #fafafa;
    --color-text-secondary: #a3a3a3;
    --root-background: linear-gradient(to top right, var(--color-background-primary), var(--color-background-primary), var(--color-background-secondary));
    --color-danger: #fca5a5;
    --color-cta-background: #a3a3a3;
    --color-cta-hover-background: #fafafa;
    --color-cta-text: #404040;
    --color-cta-hover-text: var(--color-cta-text);
    --color-cta-alt-background: #737373;
    --color-cta-alt-hover-background: #404040;
    --color-cta-alt-text: #fafafa;
    --color-cta-alt-hover-text: var(--color-cta-alt-text);
  }
}

.btn {
  height: 2em;
  border-radius: 1em;
  text-align: center;
  border: 0;
  box-shadow: var(--shadow);
  transition: all var(--transition-time);
  padding: 0 1em;
}

.btn.cta {
  background-color: var(--color-cta-background);
  color: var(--color-cta-text);
}

.btn.cta:hover,
.btn.cta:active {
  background-color: var(--color-cta-hover-background);
  color: var(--color-cta-hover-text);
}

.btn.cta.alt {
  background-color: var(--color-cta-alt-background);
  color: var(--color-cta-alt-text);
}

.btn.cta.alt:hover,
.btn.cta.alt:active {
  background-color: var(--color-cta-alt-hover-background);
  color: var(--color-cta-alt-hover-text);
}

.card {
  position: relative;
  height: 100%;
}

.card .card-header {
  display: flex;
  justify-content: space-between;
  margin-bottom: 0;
  background-color: var(--color-background-tertiary);
  padding: 1rem;
  border-bottom: solid 1px rgba(154, 161, 171, 0.15);
  border-bottom: solid 1px var(--color-border);
}

.card .card-header .header-title {
  flex: 1 1 auto;
  font-size: 1.2em;
}

.card .card-header .dropdown {
  position: relative;
  display: inline-block;
}

.card .card-header .dropdown .dropdown-btn {
  color: white;
  padding: 6px;
  font-size: 1em;
  border: none;
  cursor: pointer;
}

.card .card-header .dropdown .dropdown-content {
  position: absolute;
  min-width: 160px;
  box-shadow: 0px 8px 16px 0px rgba(0, 0, 0, 0.2);
  z-index: 1;
}

.card .card-header .dropdown .dropdown-content a {
  color: black;
  font-size: 1em;
  padding: 12px 16px;
  -webkit-text-decoration: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: block;
}

.card .card-header .dropdown .dropdown-content a:hover {
  background-color: #f1f1f1;
}

.card .card-img {
  width: 100%;
}

.card .card-body {
  padding: 5px;
}

.card-body-grid {
  display: grid;
  grid-template-columns: repeat(12, 1fr);
  grid-column-gap: 10px;
}

.card .card-body p {
  font-size: 1rem;
}

.card .card-footer {
  flex: 1 1 auto;
  padding: 5px;
  font-size: 0.75em;
  background-color: var(--color-background-tertiary);
}

.form-table {
  text-align: left;
}

.form-table label {
  font-weight: bold;
}

.form-table input,
.form-table textarea,
.form-table select {
  padding: 0.5em;
  border-radius: 0.5em;
  width: 100%;
  border: 1px solid var(--border-color);
  background: var(--color-background-primary);
  color: var(--color-text-primary)
}

.form-table .helptext {
  color: var(--color-text-secondary);
  margin: 0;
  font-size: 0.7rem;
}

.errorlist {
  color: var(--color-danger);
  list-style: none;
  text-align: left;
  padding-left: 10px;
  margin: 0;
}

.form-table .errorlist {
  padding-left: 0;
}

* {
  box-sizing: border-box;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-kerning: auto;
}

html {
  font-size: 1.2em;
  line-height: 1.4;
  font-weight: 400;
  font-family: Inter var,ui-sans-serif,system-ui, -apple-system, Segoe UI, Roboto, Ubuntu, Cantarell, Noto Sans, sans-serif,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,Noto Sans,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol,Noto Color Emoji;
  background: var(--root-background);
  color: var(--color-text-primary);
}

body {
  padding: 1em;
  margin: 0 auto;
  max-width: 95%;
  min-height: 100vh;
}

a{
  color: var(--color-text-primary);
  cursor: pointer;
  font-size: 1.1rem;
  -webkit-text-decoration: none;
  text-decoration: none;
}

h1, h2, h3, h4, h5, h6 {
  margin: 0 0 .5em 0;
  line-height: 1.2;
  letter-spacing: -.02em;
}

hr{
  border-color: var(--color-text-primary);
}

.menu {
  background: var(--color-background-secondary);
  list-style: none;
  margin: 0;
  padding-left: 0;
  display: flex;
  z-index: 2;
}

.menu li {
  background: var(--color-background-secondary);
  display: block;
  float: left;
  padding: 0.5rem 2rem;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.menu li a {
  color: var(--color-text-primary);
}

.menu li.active, .menu li ul li.active {
  background-color: var(--color-background-tertiary);
}

.menu li:hover,
.menu li:focus-within {
  background-color: var(--color-background-tertiary);
  cursor: pointer;
}

.menu li:focus-within a {
  outline: none;
}

.menu .dropdown {
  display: none;
  left: 0;
  margin: 0.5rem 0 0;
  padding: 0;
  min-width: 4em;
  opacity: 0;
  position: absolute;
  visibility: hidden;
  background: var(--color-background-tertiary);
  border: 1px solid var(--color-border);
  border-top: 0;
  overflow: hidden;
  border-radius: 0 0 15px 15px;
}

.menu li:hover > .dropdown,
.menu li:focus-within > .dropdown,
.menu li .dropdown:hover,
.menu li .dropdown:focus {
  visibility: visible;
  opacity: 1;
  display: block;
  z-index: 1;
}

.menu .dropdown li {
  clear: both;
  width: 100%;
  padding: 0;
}

.menu .dropdown li a {
  display: block;
  padding: 0.5rem 2rem;
}

.menu-card {
  display: flex;
  margin-bottom: 1em;
}

.menu-card nav {
  flex-grow: 1;
  box-shadow: var(--shadow);
  border: 1px solid var(--color-border);
  border-right: 0;
  border-left: 0;
}

.menu-card .bumper {
  width: 15px;
  overflow: hidden;
  background: var(--color-background-secondary);
  border: 1px solid var(--color-border);
  z-index: 1;
}

.menu-card .bumper:first-child {
  border-radius: 15px 0 0 15px;
  border-right: 0;
}

.menu-card .bumper:last-child {
  border-radius: 0 15px 15px 0;
  border-left: 0;
}

/* This is an example - you'd ideally use a lib or scss to compile this, add responsiveness etc */

.span-1 {
  grid-column-end: span 1 !important;
}

.span-1 {
  grid-auto-rows: min-content;
}

.span-2 {
  grid-column-end: span 2 !important;
}

.span-2 {
  grid-auto-rows: min-content;
}

.span-3 {
  grid-column-end: span 3 !important;
}

.span-3 {
  grid-auto-rows: min-content;
}

.span-4 {
  grid-column-end: span 4 !important;
}

.span-4 {
  grid-auto-rows: min-content;
}

.span-5 {
  grid-column-end: span 5 !important;
}

.span-5 {
  grid-auto-rows: min-content;
}

.span-6 {
  grid-column-end: span 6;
  grid-auto-rows: min-content;
}

.span-7 {
  grid-column-end: span 7;
  grid-auto-rows: min-content;
}

.span-8 {
  grid-column-end: span 8;
  grid-auto-rows: min-content;
}

.span-9 {
  grid-column-end: span 9;
  grid-auto-rows: min-content;
}

.span-10 {
  grid-column-end: span 10;
  grid-auto-rows: min-content;
}

.span-11 {
  grid-column-end: span 11;
  grid-auto-rows: min-content;
}

.span-12 {
  grid-column-end: span 12;
  grid-auto-rows: min-content;
}

.dashboard-container {
  display: grid;
  grid-template-columns: repeat(12, 1fr);
  grid-gap: 1rem;
  gap: 1rem;
  text-align: center;
  width: 100%;
}

h1, h2, h3, h4, p, .tab-container {
  grid-column: 1 / 13;
}

.dashboard-component {
  border-radius: 15px;
  background-color: var(--color-background-secondary);
  display: grid;
  grid-template-columns: repeat(12, 1fr);
  box-shadow: var(--shadow);
  border: 1px solid var(--color-border);
}

.dashboard-component * {
  /* allow sub nesting, by assuming not set is 12 */
  grid-column-end: span 12;
}

.tab-content div.active .dashboard-component {
  border-radius: 0 15px 15px 15px;
}

.dashboard-component-inner {
  padding: 2rem;
}

.dashboard-component-cta {
  color: var(--color-text-primary);
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.3rem;
}

.tabs {
  border-top: 2px var(--color-border) solid;
  display: flex;
  flex-wrap: wrap;
  margin-left: 0;
  padding: 35px 0 0;
  list-style: none;
}

.tabs div, .tabs .tab {
  background-color: var(--color-background-secondary);
  padding: 10px 20px;
}

.tabs div a, .tabs .tab a {
  color: var(--color-text-primary);
}

.tabs div a.active, .tabs .tab a.active {
  color: var(--color-text-primary);
  border-bottom: 1px solid var(--color-background-secondary);
}

.dataTables_filter, .dataTables_length, .dataTables_info, .dataTables_paginate {
  color: var(--color-text-primary) !important;
  font-size: 0.9rem !important;
}

.dataTables_filter, .dataTables_length, .dataTables_info, .dataTables_paginate {
  margin-bottom: 10px;
}

.dataTables_length select {
  color: var(--color-text-primary) !important;
}

.table {
  border-collapse: collapse;
  border-spacing: 0;
  width: 100%;
  font-size: 0.8rem;
}

.table thead th {
  background-color: var(--color-background-secondary);
  padding: 1rem;
  text-align: left;
  text-shadow: 1px 1px 1px var(--color-border);
}

.table tbody td {
  border: solid 1px var(--color-background-secondary);
  padding: 10px;
}

.customComponent {
  background-color: var(--color-background-primary);
  border-radius: 20px;
  padding: 10px 5px;
  margin: 10px;
  font-size: 0.8em;
}

.gauge-container {
  margin: auto;
  max-width: 65%;
}
