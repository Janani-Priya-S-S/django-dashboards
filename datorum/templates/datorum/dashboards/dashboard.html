{% extends "datorum/dashboards/base.html" %}
{% load datorum %}

{% block content_title %}
  {{ dashboard.Meta.name }}
{% endblock %}

{% block content %}
    {% render_dashboard dashboard %}
{% endblock %}