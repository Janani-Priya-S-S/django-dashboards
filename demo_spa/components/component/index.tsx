export * from "./charts"
export * from "./html"
export * from "./table"
export * from "./text"
export * from "./display"