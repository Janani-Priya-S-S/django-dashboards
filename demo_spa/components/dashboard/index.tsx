export * from "./custom"
export * from "./grid"