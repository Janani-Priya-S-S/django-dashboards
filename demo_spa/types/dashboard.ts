export enum ComponentTypes {
    Text = "Text",
    SomeCalculateText =  "SomeCalculateText",
    HTML =  "HTML",
    Plotly =  "Plotly",
    Stat =  "Stat",
    Table =  "Table",
}

export type Component = {
  key: string
  value: any
  width: number
  isDeferred: boolean
  renderType: ComponentTypes
}

export type DashboardMeta = {
  name: string,
  slug: string,
}

export type Dashboard = {
  Meta: DashboardMeta
  components: Component[]
}

