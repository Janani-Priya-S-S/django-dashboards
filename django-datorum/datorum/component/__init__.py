from .base import Component
from .chart import Chart
from .form import Form
from .map import Map
from .table import BasicTable, Table
from .text import CTA, Stat, Text


__all__ = [
    "Component",
    "CTA",
    "Chart",
    "Form",
    "Map",
    "Stat",
    "Text",
    "Table",
    "BasicTable",
]
