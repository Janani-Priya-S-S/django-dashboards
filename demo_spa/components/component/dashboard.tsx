import React, {useContext} from "react";
import {Component as ComponentType, Dashboard, DashboardComponentTypes, Value} from "@/types";
import {gql, useQuery} from "@apollo/client";
import {Stat} from "@/components/component/text";
import {Plotly} from "@/components/component/charts";
import {Tabulator} from "@/components/component/table";
import {HTML} from "@/components/component/html";
import {Form} from "@/components/component/forms";
import * as styles from "@/components/component/index.module.scss";
import {FilterContext} from "../../appContext";

/*
  We'd want something a bit more complex here/with the gql for a prod system, renderType
  is the component, but the nature of the `SomeCalculateText`/deferred ones is that they
  will often override. You can set `render_type` in be also but not much thought has gone into that yet.
 */
const DashboardComponentMap = {
    [DashboardComponentTypes.Text]: HTML,
    [DashboardComponentTypes.SomeCalculateText]: HTML,
    [DashboardComponentTypes.HTML]: HTML,
    [DashboardComponentTypes.Chart]: Plotly,
    [DashboardComponentTypes.Table]: Tabulator,
    [DashboardComponentTypes.Stat]: Stat,
    [DashboardComponentTypes.Map]: Plotly,
    [DashboardComponentTypes.Form]: Form,
}

type LazyComponentProps = {
    dashboard: Dashboard
    component: ComponentType
    Component: React.FC<{value: Value}>
}


const LazyComponent = ({dashboard, component, Component}: LazyComponentProps) => {
    const [filter] = useContext(FilterContext)
    const { loading, data } = useQuery(gql`
      query ($slug: String!, $key: String!, $filters: JSON) {
        component(slug: $slug, key: $key, filters: $filters) {
          value
        }
      }
    `,
        {
            variables: {
                slug:dashboard.Meta.slug,
                key: component.key,
                filters: filter
            }
        }
    );

    return <>
       {loading || !data ? <>Loading...</> : <Component value={data.component.value}/>}
    </>
};

type DashboardComponentProps = {
    dashboard: Dashboard
    component: ComponentType
    className?: string
}

export const DashboardComponent = ({dashboard, component}: DashboardComponentProps) => {
    const Component = DashboardComponentMap[component.renderType]
    return <div className={styles.componentInner}>
        {Object.keys(DashboardComponentMap).includes(component.renderType) ?
            !component.isDeferred ?
                <Component value={component.value}/>
            :
                <LazyComponent dashboard={dashboard} component={component} Component={Component}/>
        :
            <div>Missing dashboard component mapping</div>
        }
    </div>
}

