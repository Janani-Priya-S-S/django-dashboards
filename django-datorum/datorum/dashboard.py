import inspect
import logging
from copy import deepcopy

from django.http import HttpRequest
from django.template.loader import render_to_string
from django.utils.safestring import mark_safe

from datorum import permissions
from datorum.component import Component
from datorum.registry import registry


logger = logging.getLogger(__name__)

# todo: move to a default settings when available
DEFAULT_PERMISSION_CLASSES = [permissions.AllowAny]


class DashboardRenderMixin:
    class Layout:
        """
        Components works a like fields/formsets on contrib forms. e.g.

        components = {
            "group_one": {
                "components": ["a", "b", "c"],
                "component_widths": [3, 6],
                "group_width": 6,
            },
            "group_two": {
                "components": ["d", "e"],
                "group_width": 3,
                # no component_widths so will default to first component.width or Layout.default_component_width
            },
            # None can be uses as a catch all for the remaining non grouped ones.
            None: {
                "components": ["calculated_example", "chart_example"],
            }
        }

        """

        components: dict = {}
        default_component_width: int = 4

    def get_context(self):
        raise NotImplementedError(
            "Subclasses of DashboardRenderMixin must provide a get_context() method."
        )

    def render(self, template_name=None, layout_context=None):
        context = self.get_context()
        context.update(layout_context)
        return mark_safe(render_to_string(template_name, context))

    __str__ = render
    __html__ = render

    def as_div(self):
        """Render as <div> components."""
        return self.render("datorum/layout/div.html", {"layout": self.Layout()})

    def as_grid(self):
        """Render as <div> grid components."""
        return self.render("datorum/layout/grid.html", {"layout": self.Layout()})

    @classmethod
    def apply_layout(cls, components: list[Component]) -> list[Component]:
        """
        If a fixed layout has been defined, order, group and annotate components as needed for rendering.
        Otherwise return as per the attribute order.
        """
        layout = cls.Layout
        width = layout.default_component_width

        # if we have a defined layout, fields not in this layout are not added/shown.
        components_with_layout = []
        if layout.components:
            components_to_keys = {c.key: c for c in components}
            for group, attrs in layout.components.items():
                for i, key in enumerate(attrs.get("components")):
                    component = components_to_keys.get(key)
                    if component:
                        try:
                            component.width = attrs.get("component_widths", [])[i]
                        except IndexError:
                            logger.debug(f"Missing component_widths for {key} at {i}")
                            if not component.width:  # if not set directly on component
                                component.width = width

                        component.group_width = attrs.get("group_width")
                        component.group = attrs.get("group", group)
                        components_with_layout.append(component)
                    else:
                        logger.warning(f"Missing component for {key}")

        # default layout
        else:
            for component in components:
                if not component.width:
                    component.width = width
                components_with_layout.append(component)

        return components_with_layout


class DashboardMetaClass(type):
    def __new__(cls, clsname, bases, attrs):
        newclass = super().__new__(cls, clsname, bases, attrs)
        registry.register(newclass)
        return newclass


class Dashboard(DashboardRenderMixin, metaclass=DashboardMetaClass):
    permission_classes = DEFAULT_PERMISSION_CLASSES

    def __init__(self, request: HttpRequest = None):
        self.request = request

    def get_context(self):
        return {"components": self.get_components(), "request": self.request}

    @classmethod
    def get_attributes_order(cls):
        """
        Get the order of the attributes as they are defined on the Dashboard class.
        Follows mro, then reverses to parents first.
        """
        attributes_to_class = []
        attributes_to_class.extend(
            [list(vars(bc).keys()) for bc in cls.__mro__ if issubclass(bc, Dashboard)]
        )
        attributes_to_class.sort(reverse=True)
        return [a for nested in attributes_to_class for a in nested]

    @classmethod
    def get_components(cls, with_layout=True) -> list[Component]:
        attributes = inspect.getmembers(cls, lambda a: not (inspect.isroutine(a)))

        components_to_keys = {}
        awaiting_dependents = {}
        for key, component in attributes:
            if isinstance(component, Component):
                if not component.key:
                    component.key = key
                if not component.render_type:
                    component.render_type = component.__class__.__name__
                components_to_keys[key] = component

                if component.dependents:
                    awaiting_dependents[key] = component.dependents

        for component, dependents in awaiting_dependents.items():
            components_to_keys[component].dependent_components = [
                components_to_keys.get(d) for d in dependents  # type: ignore
            ]

        components = list(components_to_keys.values())
        components.sort(key=lambda c: cls.get_attributes_order().index(c.key))

        if with_layout:
            components = cls.apply_layout(components=deepcopy(components))

        return components

    class Meta:
        name: str
