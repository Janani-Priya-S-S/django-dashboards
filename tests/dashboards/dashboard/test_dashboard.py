import pytest

from tests.dashboards.app1.dashboards import TestDashboard, TestModelDashboard
from wildcoeus.dashboards.component import Text
from wildcoeus.dashboards.dashboard import Dashboard


pytest_plugins = [
    "tests.dashboards.fixtures",
]


def test__get_attributes_order(dashboard, snapshot):
    snapshot.assert_match(dashboard.components.keys())


def test__get_attributes_order__with_parent(complex_dashboard, snapshot):
    snapshot.assert_match(complex_dashboard.components.keys())


def test__get_components__no_layout(dashboard, rf):
    request = rf.get("/")
    assert dashboard(request=request).get_components() == [
        dashboard.components["component_1"],
        dashboard.components["component_2"],
        dashboard.components["component_3"],
    ]


def test__get_components__with_parent__no_layout(complex_dashboard, rf):
    request = rf.get("/")

    assert complex_dashboard(request=request).get_components() == [
        complex_dashboard.components["component_1"],
        complex_dashboard.components["component_2"],
        complex_dashboard.components["component_3"],
        complex_dashboard.components["component_4"],
        complex_dashboard.components["component_5"],
        complex_dashboard.components["component_6"],
        complex_dashboard.components["component_7"],
    ]


@pytest.mark.django_db
def test_dashboard__get_absolute_url(dashboard, rf):
    request = rf.get("/")
    assert dashboard(request=request).get_absolute_url() == "/app1/testdashboard/"


@pytest.mark.django_db
def test_dashboard__with_get_FOO_methods(dashboard, rf):
    class TestDashboardSingle(Dashboard):
        component_value = Text()
        component_defer = Text()

        def get_component_value_value(self):
            return "Foo"

        def get_component_defer_defer(self):
            return "Bar"

    request = rf.get("/")

    assert (
        TestDashboardSingle(request=request).components["component_value"].value()
        == "Foo"
    )
    assert (
        TestDashboardSingle(request=request).components["component_value"].defer is None
    )
    assert (
        TestDashboardSingle(request=request).components["component_defer"].defer()
        == "Bar"
    )
    assert (
        TestDashboardSingle(request=request).components["component_defer"].value is None
    )


@pytest.mark.parametrize(
    "dashboard_class",
    [
        {"klass": TestDashboard, "expected": "testdashboard"},
        {"klass": TestModelDashboard, "expected": "testmodeldashboard"},
    ],
)
def test_dashboard__class_name_method(dashboard_class):
    assert dashboard_class["klass"].class_name() == dashboard_class["expected"]


@pytest.mark.parametrize(
    "dashboard_class",
    [
        {"klass": TestDashboard, "expected": "app1_testdashboard"},
        {"klass": TestModelDashboard, "expected": "app1_testmodeldashboard"},
    ],
)
def test_dashboard__get_slug_method(dashboard_class):
    assert dashboard_class["klass"].get_slug() == dashboard_class["expected"]


def test_dashboard__str(dashboard, rf):
    request = rf.get("/")

    assert str(dashboard(request)) == "Test Dashboard"


# def test_dashboard__render__template_name(dashboard, rf, snapshot):
#     request = rf.get("/")
#     d = dashboard(request)
#     snapshot.assert_match(d.render(request, template_name="foo.html"))


# More tests to add here re layout
