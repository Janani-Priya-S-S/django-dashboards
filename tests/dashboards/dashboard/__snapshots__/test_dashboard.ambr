# serializer version: 1
# name: test__get_attributes_order
  dict_keys(['component_1', 'component_2', 'component_3'])
# ---
# name: test__get_attributes_order__with_parent
  dict_keys(['component_1', 'component_2', 'component_3', 'component_4', 'component_5', 'component_6', 'component_7'])
# ---
